snowdrop-dba-avlp/src/Snowdrop/Dba/AVLp/Accum.hs

```diff
@@ -147,19 +146,19 @@ modAVL
       , MonadCatch m
       , RetrieveImpl (ReaderT ctx m) h
       )
-    => (AVL.Map h k v, Set AVL.Revision)
+    => (AVL.Map h k v, Set h)
     -> (k, ValueOp v)
-    -> AVLCacheT h (ReaderT ctx m) (AVL.Map h k v, Set AVL.Revision)
+    -> AVLCacheT h (ReaderT ctx m) (AVL.Map h k v, Set h)
 modAVL (avl, touched) (k, valueop) = processResp =<< AVL.lookup k avl
   where
-    processResp :: ((Maybe v, Set AVL.Revision), AVL.Map h k v)
-                -> AVLCacheT h (ReaderT ctx m) (AVL.Map h k v, Set AVL.Revision)
+    processResp :: ((Maybe v, Set h), AVL.Map h k v)
+                -> AVLCacheT h (ReaderT ctx m) (AVL.Map h k v, Set h)
     processResp ((lookupRes, (<> touched) -> touched'), avl') =
       case (valueop, lookupRes) of
         (NotExisted, Nothing) -> pure (avl', touched')
-        (New v     , Nothing) -> (, touched') . snd <$> AVL.insert' k v avl'
-        (Rem       , Just _)  -> (, touched') . snd <$> AVL.delete' k avl'
-        (Upd v     , Just _)  -> (, touched') . snd <$> AVL.insert' k v avl'
+        (New v     , Nothing) -> (, touched') . snd <$> AVL.insert k v avl'
+        (Rem       , Just _)  -> (, touched') . snd <$> AVL.delete k avl'
+        (Upd v     , Just _)  -> (, touched') . snd <$> AVL.insert k v avl'
```

Comment:
Actually, this should not work. For any `k`, the trace of `insert k v` is not the same as trace of `lookup k`.

Insert can cause rebalance, and rebalance will touch additional nodes. Why does this code even do this?

I suggest replacing `(, touched') . snd` with `second (<> touched')`.

Reply:
I.e., throwing out (`snd`) the `Set h` of insert and putting `Set h` of lookup is incorrect.